controllers/rcj_soccer_referee_supervisor/referee/penalty_area_checker.py

```diff
@@ -0,0 +1,73 @@
+from typing import List
+
+from referee.consts import BLUE_PENALTY_AREA, YELLOW_PENALTY_AREA
+
+
+class PenaltyAreaChecker:
+    def __init__(self, allowed: int, reset_after: int, time: int):
+        self.allowed = allowed
+        self.reset_after = reset_after
+        self.time = time
+
+        self.y_vertical, self.y_lower, self.y_upper = YELLOW_PENALTY_AREA
+        self.b_vertical, self.b_lower, self.b_upper = BLUE_PENALTY_AREA
+        self.reset()
+
+    def reset(self):
+        self.time_entered_penalty = -1
+        self.time_left_penalty = -1
+
+    def is_in_yellow_penalty(self, x: int, z: int) -> bool:
+        return x < self.y_vertical and self.y_lower < z < self.y_upper
+
+    def is_in_blue_penalty(self, x: int, z: int) -> bool:
+        return x > self.b_vertical and self.b_lower < z < self.b_upper
+
+    @property
+    def is_outside_of_penalty_for_longer(self) -> bool:
+        return self.time < self.time_left_penalty - self.reset_after
+
+    @property
+    def is_inside_of_penalty_for_long(self) -> bool:
+        return self.time < self.time_entered_penalty - self.allowed
+
+    @property
+    def has_entered(self) -> bool:
+        return self.time_entered_penalty != -1
+
+    @property
+    def has_left(self) -> bool:
+        return self.time_left_penalty != -1
+
+    def track(self, position: List[float], time: int):
+        self.time = time
+        x, z = position[0], position[2]
+
+        if self.is_in_blue_penalty(x, z) or self.is_in_yellow_penalty(x, z):
+            # the robot enters the penalty area for the first time
+            if not self.has_entered:
+                self.time_entered_penalty = self.time
+            # the robot re-enters the penalty area
+            elif self.has_left:
+                self.time_left_penalty = -1
+        else:
+            # the robot has left the penalty area
+            if self.has_entered and not self.has_left:
+                self.time_left_penalty = self.time
+            # the robot keeps being outside the penalty area for longer
+            elif self.has_left and self.is_outside_of_penalty_for_longer:
+                self.reset()
+
+    def is_violating(self) -> int:
+        """
+        Detect whether the robot stays for longer period of time inside
+        the penalty area.
+
+        Returns: bool - whether the robot is violating this rule
+        """
+        if self.has_entered and not self.has_left:
+            if self.is_inside_of_penalty_for_long:
```

Comment:
@Adman looking at the use of this property, I am if something like `is_inside_penalty_over_limit` would not be a better name. Feel free to ignore this though.